### PR TITLE
fix(replays): Add handling for upload errors in buffered consumer

### PR DESCRIPTION
**Pull request by cmanallen**
Adds handling for errors while uploading to the service-provider.  Specifically it:

1. Captures and logs errors that occur inside the threadpool.
2. Raises an exception which resets the buffer and begins processing from the previously committed offset.

This was added to address (in a safe and reliable way) the "BadGateway" error that GCS throws.

### CODE REVIEW

**Review by JoshFerge — COMMENTED** [review]
i forget, does our uploading have any retry logic? what happens now when an upload to our service provider fails in the non buffered consumer?

**Comment by cmanallen** [comment]
@JoshFerge Re: current behavior.  Its supposed to bubble up exceptions from the threadpool.  Which should reset us back to the previously committed offset.  That's something I can verify (and fix if need be).

Re: upload retry.  I believe we do yes.  To the best of my knowledge its 5 retries and its defined somewhere in Sentry.  Another thing I can verify for you :D

**Comment by JoshFerge** [comment]
> @JoshFerge Re: current behavior. Its supposed to bubble up exceptions from the threadpool. Which should reset us back to the previously committed offset. That's something I can verify (and fix if need be).
> 
> Re: upload retry. I believe we do yes. To the best of my knowledge its 5 retries and its defined somewhere in Sentry. Another thing I can verify for you :D

just curious if it crashes the consumer now. i'm curious if crashing the consumer here will result in any negative consequences. --

seems like now one crash will result in many things being retried,  so a small % of failures from GCS seems like it would cascade and we could end up in a near ending retry loop even if only ~1-2% uploads are failing?

**Comment by cmanallen** [comment]
@JoshFerge 

- GCS retry behavior is documented [here](https://github.com/getsentry/sentry/blob/3ef9e6a977fa2e4a4d5b877ce993f0b0651a4c81/src/sentry/filestore/gcs.py#L39-L74).
    - 5 retries by default.
- Arroyo [will bubble up exceptions](https://github.com/getsentry/sentry/blob/3ef9e6a977fa2e4a4d5b877ce993f0b0651a4c81/.venv/lib/python3.10/site-packages/arroyo/processing/strategies/run_task_in_threads.py#L105-L111) raised in the worker.  If unhandled this will crash the process and force a restart from the previously committed offset.
    - Its important to note that deterministic, unhandled exceptions will block any consumer implementation indefinitely.
    - Intermittent exceptions like we observed with the service-provider faults will block the consumer until the fault is resolved.
    - The dollar cost of an exception depends on the throughput of the consumer.  Not the number of workers running in parallel.
        - The dollar cost can be calculated as "number of messages processed since the last committed offset" multiplied by the "cost per message".
    - The new consumer implementation has the potential to increase the cost of service-provider outages because it increases the throughput of the consumer.
    - This could be offset by committing more frequently (after each batch as an example).
    - However, the cost per message is more significantly impacted by the number-of and size-of our consumers.  An increase in throughput more than offsets the increase in GCS cost.
- If the consumer experiences one intermittent exception per second the consumer will deadlock regardless of implementation.
    - The throughput of the consumer determines if we deadlock.  If our throughput is slow enough that we experience fewer than one exception per second then we will make progress (slowly).
    - On the bright side we don't experience an exception rate at that level.  As best I can tell from the Sentry data we have eight 9's of reliability.  If we encountered errors at 10,000x the rate we should still be able to make progress.

**Comment by cmanallen** [comment]
Added a docstring so we have a long-lived record of this information 👍

File: src/sentry/replays/consumers/recording_buffered.py [hunk 1]
```diff
@@ -275,8 +312,27 @@ def process_message(buffer: RecordingBuffer, message: Message[KafkaPayload]) ->
...
         with ThreadPoolExecutor(max_workers=len(upload_events)) as pool:
-            pool.map(_do_upload, upload_events)
+            futures = [pool.submit(_do_upload, upload) for upload in upload_events]
+
+    has_errors = False
+
+    # These futures should never fail unless there is a service-provider issue.
+    for error in filter(lambda n: n is not None, (fut.exception() for fut in futures)):
+        has_errors = True
+        sentry_sdk.capture_exception(error)
```
Comment (untitaker) [inline src/sentry/replays/consumers/recording_buffered.py:326]: why not `raise error` right here?

Comment (cmanallen) [reply src/sentry/replays/consumers/recording_buffered.py:326]: Trying to capture all of the errors not just one of them.  I could raise the last error.

**Review by untitaker — APPROVED** [review]

**Review by JoshFerge — APPROVED** [review]

**Comment by JoshFerge** [comment]
could always try to increase retry limit / use exponential backoffs if intermittent uploads ever cause problems. thank you for the explanation @cmanallen!

**Comment by cmanallen** [comment]
@JoshFerge That's a good point.  I checked out `try_repeated` (the function that manages GCS retries) and it does not have any back-off behavior built-in!  It's definitely something we should consider.

**Pull request merged**